linera-chain/src/chain.rs

```diff
@@ -541,12 +558,8 @@ where
             message_counts
                 .push(u32::try_from(messages.len()).map_err(|_| ArithmeticError::Overflow)?);
         }
-        let used_fuel = available_fuel.saturating_sub(remaining_fuel);
-
         let balance = self.execution_state.system.balance.get_mut();
-        Self::sub_assign_fees(balance, credit)?;
-        Self::sub_assign_fees(balance, pricing.fuel_price(used_fuel)?)?;
-        Self::sub_assign_fees(balance, pricing.messages_price(&messages)?)?;
+        sub_assign_fees(balance, credit)?;
```

Comment:
I am not sure what you mean by "added twice", but the logic is definitely complicated and took me some time to deal with. But I think this should be in separate PRs.

Reply:
`credit` is the sum of all incoming transfers in this block. To avoid failing due to insufficient balance we add this to the balance before subtracting any costs. That way, even if your current balance is 0, if you have an incoming transfer you can create a new block and receive that transfer.

Since during the actual execution of the block the transfer messages are processed, their amounts are added to the chain balance _again_. That's why, after execution, we subtract `credit` again.

With this PR, however, we use the intermediate values of `balance` to compute our fuel limit, too, and in these, the incoming transfers will sometimes already be counted twice.

To fix this, instead of subtracting `credit` again at the end, we could make the execution of a transfer a no-op. Not really elegant either, though. :confused:

Reply:
Did the internal discussion clarify this? or do we need to coordinate?

Reply:
To summarize the discussion:
* Let's remove the `credit` thing: If you have 0 balance, you have to make sure that the first incoming message is a transfer that covers the block's fees.
* Subtract the base fee for the block _after_ processing messages and operations.
* Remove the fee for storing incoming messages. Consider this to be included in the `messages_price` on the sender chain.

Reply:
Ok, I would propose to do that after in another PR.

Reply:
I created the corresponding issue https://github.com/linera-io/linera-protocol/issues/1187